doc/design/program.md

```diff
@@ -120,6 +120,22 @@ message AttrDesc {
 }
 ```
 
+## ProgramDesc and ExecutionPlan
+
+The goal of `ProgramDesc` is to describe **what** the user wants to calculate, and the goal of `ExecutionPlan` is to specify **how** to calculate it.
+
+For example, the `ExecutionPlan` has OP placement information to indicate which device the OP will run, but the `ProgramDesc` does not have this information since currently our Python API does not support manually pinning an OP onto a type of device (e.g., GPU or FPGA). On the other hand, the `ProgramDesc` should have information about if an OP belongs to an optimizer, this information is provided by the user and helps to place the OPs onto the parameter servers, but the `ExecutionPlan` does not have this information.
+
+### Optimizer
+
+The optimizer takes `ProgramDesc` as the input and outputs the `ExcutionPlan`, the steps are:
+1. Add the prgram in `ProgramDesc` and the coresponding backward pass program into the `ExecutionPlan`.
+1. Optimizes the program according to the avaiable devices.
```

Comment:
I am a little confused at `the avaiable devices`. 
Which part should own the `Optimizer` module? The cluster or the client program? 

Especially in the Elastic DeepLearning, if the user request for nodes in a range 5-10, how should we generate the `ExecutionPlan`?

Reply:
`Optimizer` module (not a good name, maybe assembler/transformer/? would be better) should be in a binary running in the cluster for distributed training. For local training, the module should be compiled locally.